### ISSUE TITLE
[Compatibility Version] alphas with emulated version

### ISSUE THREAD

**Issue by Jefftree** [kubernetes/kubernetes#128502]
Per the compatibility version KEP, alphas are outside the scope of compatibility version.

https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/4330-compatibility-versions/README.md#non-goals

> Support --emulation-version for Alpha features. Alpha feature are not designed to be upgradable, so we will not allow alpha features to be enabled when --emulation-version is set.

We current don't have proper safeguards for this.

- [ ] Alpha features should not be permitted to be enabled with emulated version
- [ ] Alpha APIs should not be permitted to be enabled with emulated version

/cc @aaron-prindle @jpbetz 

/triage accepted
/sig architecture
/sig api-machinery

**Comment by siyuanfoundation** [comment]
I don't think we necessarily need to break the server if alpha features/apis are enabled in emulation mode. We do not officially support this use case, but it does not mean it will always break. If a user wants to use a alpha feature that has not changed for a while it would work fine. Users can use it at their own risk. We can signal that through warning msg.

Detailed instructions of how to implement this:

- Print warning if any Alpha features are enabled with emulated version
  - through the [`featuregate.GetAll()`](https://github.com/kubernetes/kubernetes/blob/354299bea07760f83a0aeaa2023cbd0319f3cad0/staging/src/k8s.io/component-base/featuregate/feature_gate.go#L517) interface you can get all the features and their stage at the emulation version.
  - in [`func (r *componentGlobalsRegistry) Validate()`](https://github.com/kubernetes/kubernetes/blob/354299bea07760f83a0aeaa2023cbd0319f3cad0/staging/src/k8s.io/component-base/compatibility/registry.go#L381) check if any alpha feature is enabled when emulation version != binary version, and print out warnings
- Print warning if any Alpha apis are enabled with emulated version
  - `--runtime-config` of APIs are applied at [`func (s *APIEnablementOptions) ApplyTo`](https://github.com/kubernetes/kubernetes/blob/354299bea07760f83a0aeaa2023cbd0319f3cad0/staging/src/k8s.io/apiserver/pkg/server/options/api_enablement.go#L92), we can check if any alpha APIs are enabled at that point if `c.EffectiveVersion().BinaryVersion() != c.EffectiveVersion().EmulationVersion() `

**Comment by henrybear327** [comment]
PR is almost ready, fighting some unit test failures now still :( 

Sorry for the delay.

**Comment by henrybear327** [comment]
As this is a high priority task and I can't manage to fix the test in time, probably better for @michaelasp to take over :(

/unassign @henrybear327 
/assign @michaelasp

**Comment by michaelasp** [comment]
/assign @michaelasp 

Let me take a look at this
